File: LeetCode/Explore/TopInterview/Blind75/DynamicProgramming/139_word_break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
      dp = [False] * (len(s) + 1)
      dp[0] = True
      
      sp = set(wordDict)
      
      for i in range(1, len(s) + 1, 1):
        for j in range(0, i, 1):
          if dp[j] and s[j:i] in sp:
            dp[i] = True
            break
            
      return dp[len(s)]

File: LeetCode/Explore/TopInterview/Meta/DynamicProgramming/139_word_break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
      set<string> sp(wordDict.begin(), wordDict.end());
      vector<bool> dp(s.size() + 1, false);
      dp[0] = true;
      for (int i = 1; i < s.size() + 1; i++) {
        for (int j = 0; j < i; j++) {
          if (dp[j] && sp.count(s.substr(j, i - j))) {
            dp[i] = true;
            break;
          }
        }
      }
      return dp[s.size()];
    }
};
